src/librustc_metadata/decoder.rs

```diff
@@ -465,15 +467,22 @@ crate fn proc_macro_def_path_table(crate_root: &CrateRoot<'_>,
     }
 
     definitions.def_path_table().clone()
-}
+}*/
 
 impl<'a, 'tcx> CrateMetadata {
-    fn is_proc_macro(&self, id: DefIndex) -> bool {
-        self.proc_macros.is_some() && id != CRATE_DEF_INDEX
+    fn is_proc_macro_crate(&self) -> bool {
+        self.root.proc_macro_decls_static.is_some()
+    }
+    fn is_proc_macro(&self, id: DefIndex, sess: &Session) -> bool {
+        self.is_proc_macro_crate() && self.get_proc_macros(sess).proc_macros.as_ref().unwrap()
+            .contains_key(&id)
```

Comment:
We *really* shouldn't need to check `sess` for this. In fact, I don't see why this would be correct, or where `get_proc_macros` is even coming from!

IMO this should check for `self.entry(index).kind.def_kind() == DefKind::Fn` with one of the proc macro attributes (or you could even add `proc_macro: Option<MacroKind>` the to `schema::FnData` or w/e).